lib/useOnyx.ts

```diff
@@ -0,0 +1,200 @@
+            );
+        }
+    }, [previousKey, key]);
+
+    /**
+     * According to React docs, `getSnapshot` is a function that returns a snapshot of the data in the store that’s needed by the component.
+     * **While the store has not changed, repeated calls to getSnapshot must return the same value.**
+     * If the store changes and the returned value is different (as compared by Object.is), React re-renders the component.
+     */
+    const getSnapshot = useCallback(() => {
+        /**
+         * Case 1 - We have a normal key without selector
+         *
+         * We just return the data from the Onyx cache.
+         */
+        if (!Onyx.isCollectionKey(key) && !options?.selector) {
+            return getCachedValue(key) as TReturnData;
+        }
+
+        /**
+         * Case 2 - We have a normal key with selector
+         *
+         * Since selected data is not directly stored in the cache, we need to generate it with `getCachedValue`
+         * and deep compare with our previous internal data.
+         *
+         * If they are not equal, we update the internal data and return it.
+         *
+         * If they are equal, we just return the previous internal data.
+         */
+        if (!Onyx.isCollectionKey(key) && options?.selector) {
+            const newData = getCachedValue(key, options.selector);
+            if (!deepEqual(previousDataRef.current, newData)) {
+                previousDataRef.current = newData as TReturnData;
+            }
+
+            return previousDataRef.current as TReturnData;
+        }
+
+        /**
+         * Case 3 - We have a collection key with/without selector
+         *
+         * Since both collection objects and selected data are not directly stored in the cache, we need to generate them with `getCachedValue`
+         * and deep compare with our previous internal data.
+         *
+         * If they are not equal, we update the internal data and return it.
+         *
+         * If they are equal, we just return the previous internal data.
+         */
+        const newData = getCachedValue(key, options?.selector);
+        if (!deepEqual(previousDataRef.current, newData)) {
+            previousDataRef.current = newData as TReturnData;
+        }
+
+        return previousDataRef.current as TReturnData;
+    }, [key, options?.selector]);
+
+    const subscribe = useCallback(
+        (onStoreChange: () => void) => {
+            connectionIDRef.current = Onyx.connect({
+                key: key as CollectionKeyBase,
+                callback: () => {
+                    /**
+                     * We don't need to update the Onyx cache again here, when `callback` is called the cache is already
+                     * expected to be updated, so we just signal that the store changed and `getSnapshot()` can be called.
+                     */
+                    fetchStatusRef.current = 'loaded';
+                    onStoreChange();
```

Comment:
Does it matter if we swap these?
```suggestion
                    onStoreChange();
                    fetchStatusRef.current = 'loaded';
```